### ISSUE TITLE
`rfd` file dialogs crashes `egui_web`

### ISSUE THREAD

**Issue by rnd-ash** [emilk/egui#1042]
**Describe the bug**
Tested on both master and 0.16.1 versions of EGUI, with RFD 0.6.3 (Latest), calling `block_on` an async file dialog in  update loop for egui causes constant crashing if the file dialog is open.

Async file dialog is required in web as described in RFD documentation.

This behavior is not present when running on native target (Async file dialog works fine), only on Webasm target.

The crash seems to happen in `egui_web::install_document_events::{{closure}}`

Code
```rust

// Returns byte array of DAT file contents
async fn read_open_file(ctx: &epi::Frame) -> Option<Vec<u8>> {
    let file = rfd::AsyncFileDialog::new()
        .add_filter("DAT", &["dat"])
        .pick_file().await;
    if let Some(f) = file { // File is chosen. Read the bytes
        return Some(f.read().await)
    } else { // No file chosen
        return None;
    }
}
...

fn update(&mut self, ctx: &egui::CtxRef, frame: &epi::Frame) {
    egui::CentralPanel::default().show(ctx, |ui|{
        if ui.button("Select File").clicked() {
          if let Some(buffer) = block_on(read_open_file(frame)) {
              // TODO Manage opened file contents
          }
       }
    }
}
```

**Screenshots of crash + stack trace**
![image](https://user-images.githubusercontent.com/14551965/148210926-a54dbd95-f84d-4f97-a6a5-103ffdb1516f.png)

**Comment by TENX-S** [comment]
I don't know if this has anything to do with the error you're getting, but it's not a good idea to use block_on in `update` because it blocks the UI from redrawing, so it can't respond to any events.

**Comment by emilk** [comment]
Have you gotten `rfd` to work on web without using `eframe`?

**Comment by rnd-ash** [comment]
how so? I thought eframe is required (Building an app that runs on web + native)

**Comment by emilk** [comment]
e.g. using [`egui-miniquad`](https://github.com/not-fl3/egui-miniquad). I'm not suggestion you switch from eframe, just help test and try to pin-point why rdf fails. Perhaps this is an issue in `rdf` rather than in `egui_web`?

**Comment by emilk** [comment]
I think the problem is that you are using `block_on`. You should probably be using `wasm_bindgen_futures::spawn_local` instead, as shown in the `rfd` example: https://github.com/PolyMeilex/rfd/blob/master/examples/async.rs

**Comment by margual56** [comment]
> You should probably be using `wasm_bindgen_futures::spawn_local`

Yes, but then you have no way of getting the information outside the other thread. 

If I want to read a file in WASM, the only way is to use an async function:
```Rust
async fn read_open_file() -> Option<Vec<u8>> {
    let file = rfd::AsyncFileDialog::new().pick_file().await;
    if let Some(f) = file { // File is chosen. Read the bytes
        return Some(f.read().await)
    } else { // No file chosen
        return None;
    }
}
```

And if I then call that function from the UI:
```Rust
egui::TopBottomPanel::top("header").show(ctx, |ui| {
  if ui.button("TEST READ FROM WASM").clicked() {
    let data = read_open_file().await;
  }
});
```

And I get the error "`await` is only allowed inside `async` functions and blocks". Thus, I would have to:
```Rust
egui::TopBottomPanel::top("header").show(ctx, async |ui| {
  if ui.button("TEST READ FROM WASM").clicked() {
    let data = read_open_file().await;
  }
});
```
which is not possible currently in egui/eframe.

---
## TL;DR

The current implementation of egui/eframe does not support async, you also cannot block the main thread (the browser will thing the page crashed) and there is no way of getting data from a file in WASM without awaiting.

---
# Help!

I have been scouring the internet in search of an answer to this very question of **how do you get information from an async block in WASM when you have a main thread that cannot be blocked**, with no avail.

Has @rnd-ash found the answer? Did emilk's answer work for you?

**I want/need to be able to load/download files from WASM!**

## PS

You will probably try and point me to [this discussion](https://github.com/emilk/egui/discussions/2010) in the emilk discussion forum. Although the last comment is marked as an answer, it did not work for the author of the question nor did it work for me.

Why? **In my case** it does not work because it uses `poll_promise::Promise::spawn_thread`, for which you need to enable the `tokio` feature for the `poll_promise` crate, and that gives me the error:
```Rust
index.crates.io-6f17d22bba15001f/tokio-1.29.1/src/lib.rs:488:1
    |
488 | compile_error!("Only features sync,macros,io-util,rt,time are supported on wasm.");
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Comment by emilk** [comment]
> how do you get information from an async block in WASM when you have a main thread that cannot be blocked

Use a channel?

> Why? In my case it does not work because it uses `poll_promise::Promise::spawn_thread`, for which you need to enable the `tokio` feature for the `poll_promise` crate, and that gives me the error

Read the docs: https://docs.rs/poll-promise/latest/poll_promise/#web
